File: aat-android/src/main/java/ch/bailu/aat/dispatcher/EditorOrBackupSource.java
package ch.bailu.aat.dispatcher;

import ch.bailu.aat_lib.service.editor.EditorInterface;
import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.dispatcher.ContentSource;
import ch.bailu.aat_lib.dispatcher.OnContentUpdatedInterface;
import ch.bailu.aat_lib.gpx.GpxFileWrapper;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.GpxList;
import ch.bailu.aat_lib.gpx.InfoID;
import ch.bailu.foc.Foc;

public class EditorOrBackupSource extends ContentSource implements EditorSourceInterface {

    private final EditorSource editorSource;
    private final ContentSource backupSource;

    private boolean isEditing = false;


    public EditorOrBackupSource(AppContext appContext, ContentSource source) {
        editorSource = new EditorSource(appContext);
        backupSource = source;

    }

    public boolean isModified() {
        return isEditing && getEditor().isModified();
    }

    public void releaseEditorSave() {
        if (isEditing) {
            getEditor().save();
            releaseEditorDiscard();
        }
    }


    @Override
    public void setTarget(OnContentUpdatedInterface t) {
        super.setTarget(t);
        editorSource.setTarget(t);
        backupSource.setTarget(t);
    }


    public void releaseEditorDiscard() {
        if (isEditing) {
            isEditing = false;

            editorSource.onPause();
            backupSource.onResume();

            requestUpdate();
        }

    }

    public void edit() {
        Foc file = backupSource.getInfo().getFile();
        if (file != null && !isEditing) {
            isEditing = true;

            editorSource.edit(file);
            editorSource.onResume();
            backupSource.onPause();

            requestUpdate();
        }
    }


    public boolean isEditing() {
        return isEditing;
    }

    @Override
    public EditorInterface getEditor() {
        return editorSource.getEditor();
    }

    @Override
    public int getIID() {
        return editorSource.getIID();
    }

    @Override
    public GpxInformation getInfo() {
        if (isEditing) return editorSource.getInfo();
        return backupSource.getInfo();
    }


    @Override
    public void requestUpdate() {
        if (isEditing) {
            requestBackupNullUpdate();
            editorSource.requestUpdate();


        }  else {
            requestEditorNullUpdate();
            backupSource.requestUpdate();

        }
    }



    @Override
    public void onPause() {
        if (isEditing) editorSource.onPause();
        else backupSource.onPause();
    }


    @Override
    public void onResume() {
        if (isEditing) {
            requestBackupNullUpdate();
            editorSource.onResume();

        } else {
            requestEditorNullUpdate();
            backupSource.onResume();

        }
    }


    @Override
    public Foc getFile() {
        if (isEditing) return editorSource.getFile();
        return backupSource.getInfo().getFile();
    }

    private void requestEditorNullUpdate() {
        sendUpdate(InfoID.EDITOR_OVERLAY, new GpxFileWrapper(getFile(), GpxList.NULL_ROUTE));
    }

    private void requestBackupNullUpdate() {
        sendUpdate(backupSource.getIID(), new GpxFileWrapper(getFile(), GpxList.NULL_ROUTE));
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/EditorOrBackupSource.kt
package ch.bailu.aat_lib.dispatcher

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.dispatcher.usage.UsageTracker
import ch.bailu.aat_lib.dispatcher.usage.UsageTrackerAlwaysEnabled
import ch.bailu.aat_lib.gpx.information.GpxFileWrapper
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.gpx.GpxList
import ch.bailu.aat_lib.service.editor.EditorInterface
import ch.bailu.foc.Foc


class EditorOrBackupSource(appContext: AppContext, source: SourceInterface) :
    EditorSourceInterface, SourceInterface {
    private val editorSource: EditorSource = EditorSource(appContext, UsageTrackerAlwaysEnabled())
    private val backupSource: SourceInterface = source
    override var isEditing = false
        private set

    val isModified: Boolean
        get() = isEditing && editor.isModified()

    fun releaseEditorSave() {
        if (isEditing) {
            editor.save()
            releaseEditorDiscard()
        }
    }

    fun releaseEditorDiscard() {
        if (isEditing) {
            isEditing = false
            editorSource.onPauseWithService()
            backupSource.onResumeWithService()
            requestUpdate()
        }
    }

    override fun onDestroy() {}

    override fun setTarget( target: TargetInterface) {
        editorSource.setTarget(target)
        backupSource.setTarget(target)
    }

    override fun edit() {
        val file = backupSource.info.getFile()
        if (!isEditing) {
            isEditing = true
            editorSource.edit(file)
            editorSource.onResumeWithService()
            backupSource.onPauseWithService()
            requestUpdate()
        }
    }


    override val editor: EditorInterface
        get() =  editorSource.editor

    override fun getIID(): Int {
        return editorSource.getIID()
    }

    override fun getInfo(): GpxInformation {
        return if (isEditing) editorSource.info else backupSource.info
    }

    override fun requestUpdate() {
        if (isEditing) {
            requestNullUpdate(backupSource)
            editorSource.requestUpdate()
        } else {
            requestNullUpdate(editorSource)
            backupSource.requestUpdate()
        }
    }

    override fun onPauseWithService() {
        if (isEditing) editorSource.onPauseWithService() else backupSource.onPauseWithService()
    }

    override fun onResumeWithService() {
        if (isEditing) {
            requestNullUpdate(backupSource)
            editorSource.onResumeWithService()
        } else {
            requestNullUpdate(editorSource)
            backupSource.onResumeWithService()
        }
    }

    override val file: Foc
        get() =  if (isEditing) editorSource.file else backupSource.info.getFile()

    private fun requestNullUpdate(source: SourceInterface) {
        editorSource.sendUpdate(source.getIID(), GpxFileWrapper(file, GpxList.NULL_ROUTE))
    }
}
